gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java

```diff
@@ -133,9 +136,21 @@ public Void call()
             }
             generateCommitSequenceBuilder(this.datasetState, entry.getValue());
           } else {
-            DataPublisher publisher = taskFactory == null ? closer
-                .register(DataPublisher.getInstance(dataPublisherClass, this.jobContext.getJobState()))
-                : taskFactory.createDataPublisher(this.datasetState);
+            DataPublisher publisher;
+
+            if (taskFactory == null) {
+              publisher = DataPublisherFactory.get(dataPublisherClass.getName(), this.jobContext.getJobState(),
+                  this.jobContext.getJobBroker());
+
+              // non-threadsafe publishers are not shareable and are not retained in the broker, so register them with
+              // the closer
+              if (!publisher.supportsCapability(Capability.THREADSAFE, Collections.EMPTY_MAP)) {
+                closer.register(publisher);
+              }
+            } else {
+              publisher = taskFactory.createDataPublisher(this.datasetState);
```

Comment:
Please add a TODO to remember we may have to solve the shared publisher issue for task factories.